pkg/operator/csidriveroperator/crcontroller.go

```diff
@@ -0,0 +1,268 @@
+	versionName             = "CSIDriverOperator"
+)
+
+var (
+	opScheme = runtime.NewScheme()
+	opCodecs = serializer.NewCodecFactory(opScheme)
+)
+
+func init() {
+	if err := operatorapi.AddToScheme(opScheme); err != nil {
+		panic(err)
+	}
+}
+
+func NewCSIDriverOperatorCRController(
+	name string,
+	clients *csoclients.Clients,
+	csiOperatorConfig csioperatorclient.CSIOperatorConfig,
+	eventRecorder events.Recorder,
+	resyncInterval time.Duration,
+) factory.Controller {
+	f := factory.New()
+	f = f.ResyncEvery(resyncInterval)
+	f = f.WithSyncDegradedOnError(clients.OperatorClient)
+	// Necessary to do initial Sync after the controller starts.
+	f = f.WithPostStartHooks(initalSync)
+	// Add informers to the factory now, but the actual event handlers
+	// are added later in CSIDriverOperatorCRController.Run(),
+	// when we're 100% sure the controller is going to start (because it
+	// depends on the platform).
+	// If we added the event handlers now, all events would pile up in the
+	// controller queue, without anything reading it.
+	f = f.WithInformers(
+		clients.OperatorClient.Informer(),
+		clients.OperatorInformers.Operator().V1().ClusterCSIDrivers().Informer(),
+		clients.KubeInformers.InformersFor(csoclients.CSIOperatorNamespace).Apps().V1().Deployments().Informer())
+
+	c := &CSIDriverOperatorCRController{
+		name:                   name,
+		operatorClient:         clients.OperatorClient,
+		kubeClient:             clients.KubeClient,
+		operatorClientSet:      clients.OperatorClientSet,
+		clusterCSIDriverLister: clients.OperatorInformers.Operator().V1().ClusterCSIDrivers().Lister(),
+		eventRecorder:          eventRecorder.WithComponentSuffix(name),
+		factory:                f,
+		csiDriverName:          csiOperatorConfig.CSIDriverName,
+		csiDriverAsset:         csiOperatorConfig.CRAsset,
+		optional:               csiOperatorConfig.Optional,
+	}
+	return c
+}
+
+func (c *CSIDriverOperatorCRController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("CSIDriverOperatorCRController sync started")
+	defer klog.V(4).Infof("CSIDriverOperatorCRController sync finished")
+
+	var errs []error
+	opSpec, opStatus, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	// Sync CSIDriver CR
+	requiredCR := c.getRequestedClusterCSIDriver(opSpec.LogLevel)
+	var lastGeneration int64
+	lastGenerationStatus := resourcemerge.GenerationFor(
+		opStatus.Generations,
+		schema.GroupResource{Group: operatorapi.GroupName, Resource: "clustercsidrivers"},
+		"",
+		requiredCR.Name)
+	if lastGenerationStatus != nil {
+		lastGeneration = lastGenerationStatus.LastGeneration
+	}
+
+	cr, _, err := c.applyClusterCSIDriver(requiredCR, lastGeneration)
+	if err != nil {
+		klog.V(4).Infof("CSIDriver Apply failed: %s", err)
+		// This will set Degraded condition
+		return err
+	}
+	klog.V(4).Infof("CSIDriver Apply succeeded")
```

Comment:
logging and returning err & success message

Reply:
Fixed